### PR TITLE
Added a wrapper for std::bind

### PR DESCRIPTION
**Pull request by chleh**
Follow-up of #1309.

Example:
```{.cpp}

using std::placeholders;
Object some_object;

auto f_bind = std::function<ReturnType(Arg1, Arg2, Arg3)>(
                  std::bind(&Object::methodWithThreeArguments,
                      std::ref(some_object), _1, _2, _3));

auto f_easy = easyBind(&Object::methodWithThreeArguments, some_object);
```

The lines generating `f_bind` and `f_easy` produce equivalent results.

Features:
* less typing
* result type can be used for type deduction (I'll use that in a follow-up PR)
* no accidental copies of the object being bound to.
* easyBind guarantees that all types involved match up.

### CODE REVIEW

**Comment by chleh** [comment]
Jenkins, test this please.

File: BaseLib/Functional.h [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+ *
+ * The result of this function can be used, e.g., to deduce the signature of the
+ * \c method (which is not possible with the result of std::bind).
+ *
+ * Example:
+ * \code{.cpp}
+ * using std::placeholders;
+ * Object some_object;
+ *
+ * auto f_bind = std::function<ReturnType(Arg1, Arg2, Arg3>(
```
Comment (TomFischer) [inline BaseLib/Functional.h]: Closing parenthesis is missing.

File: BaseLib/TMPUtil.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+};
+
+//! Generates an IntegerSequence.
+//!
+//! \see http://stackoverflow.com/a/7858971
+template <int N, int... S>
+struct GenerateIntegerSequence {
+    // effectively pushes N-1 from the left to the list int... S of integers.
+    typedef typename GenerateIntegerSequence<N - 1, N - 1, S...>::type type;
+};
```
Comment (TomFischer) [inline BaseLib/TMPUtil.h]: Why not `std::integer_sequence` from C++14?  This is one more good reason to switch to C++14.

File: Tests/BaseLib/TestFunctional.cpp [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+    InstanceCounter() {
+        ++_num_constructed;
+    }
+    InstanceCounter(InstanceCounter const&) {
+        ++_num_copied;
+    }
+    InstanceCounter(InstanceCounter&&) {
+        ++_num_moved;
+    }
+    ~InstanceCounter() {
```
Comment (TomFischer) [inline Tests/BaseLib/TestFunctional.cpp]: Should not be the destructor virtual because you derive class `A` from `InstanceCounter`?

File: Tests/BaseLib/TestFunctional.cpp [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+    EXPECT_EQ((num_const), InstanceCounter::getNumberOfConstructions());    \
+    EXPECT_EQ((num_move), InstanceCounter::getNumberOfMoves());             \
+    EXPECT_EQ((num_copy), InstanceCounter::getNumberOfCopies());            \
+    EXPECT_EQ((num_inst), InstanceCounter::getNumberOfInstances())
+
+#define UPDATE_INSTANCES(num_const, num_move, num_copy, num_inst) \
+    num_const = InstanceCounter::getNumberOfConstructions();              \
+    num_move = InstanceCounter::getNumberOfMoves();                       \
+    num_copy = InstanceCounter::getNumberOfCopies();                      \
+    num_inst = InstanceCounter::getNumberOfInstances()
```
Comment (TomFischer) [inline Tests/BaseLib/TestFunctional.cpp]: Why do you use macros? Functions are easily possible.

File: Tests/MathLib/TestLinearSolver.cpp [hunk 5]
```diff
@@ -228,12 +228,11 @@ TEST(Math, CheckInterface_EigenLis)
 {
     // set solver options using Boost property tree
     boost::property_tree::ptree t_root;
-    boost::property_tree::ptree t_solver;
     t_root.put("lis", "-i cg -p none -tol 1e-15 -maxiter 1000");
     BaseLib::ConfigTree conf(t_root, "",
         BaseLib::ConfigTree::onerror, BaseLib::ConfigTree::onwarning);
 
-    using IntType = MathLib::LisMatrix::IndexType;
+    using IntType = MathLib::EigenMatrix::IndexType;
```
Comment (TomFischer) [inline Tests/MathLib/TestLinearSolver.cpp]: How is this related to `easyBind` PR?

File: BaseLib/Functional.h [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+ * easyBind: In easyBind \c obj is never copied, instead it will be referenced.
+ * This is in contrast to the behaviour of std::bind, and has been chosen in
+ * order to prevent accidental copies.
+ */
+template <typename Object, typename MethodClass, typename ReturnType,
+          typename... Args>
+typename std::enable_if<
+    std::is_same<MethodClass,
+                 typename std::remove_cv<typename std::remove_pointer<
+                     typename std::decay<Object>::type>::type>::type>::value,
```
Comment (TomFischer) [inline BaseLib/Functional.h]: It is not clear to me why `remove_cv` and ` remove_pointer` is needed after `decay`. Could you explain, please?

File: BaseLib/Functional.cpp [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+namespace BaseLib
+{
+namespace detail
+{
+#define DEFINE_INDEXEDPLACEHOLDER_MEMBER(INDEX, INDEX_P_1) \
+    const decltype(std::placeholders::_##INDEX_P_1)        \
+        IndexedPlacedPlaceholder<(INDEX)>::value =         \
+            std::placeholders::_##INDEX_P_1
+
+DEFINE_INDEXEDPLACEHOLDER_MEMBER(0, 1);
```
Comment (TomFischer) [inline BaseLib/Functional.cpp]: Why is the shift necessary (0->1, 1->2, ...)?

File: Tests/BaseLib/TestFunctional.cpp (hunk 4 above)
Comment (chleh) [inline Tests/BaseLib/TestFunctional.cpp]: the above must be a macro in order to get the right line in gtest's messages. I'll change the method below.

File: Tests/MathLib/TestLinearSolver.cpp (hunk 5 above)
Comment (chleh) [inline Tests/MathLib/TestLinearSolver.cpp]: It's already contained in the (by now merged) Lis PR.

File: BaseLib/Functional.h (hunk 6 above)
Comment (chleh) [inline BaseLib/Functional.h]: At some point I found it to be necessary. I also wondered. I can check it again.

File: BaseLib/Functional.cpp (hunk 7 above)
Comment (chleh) [inline BaseLib/Functional.cpp]: std::placeholders start with `_1`, but the index starts with 0.

File: Tests/BaseLib/TestFunctional.cpp (hunk 3 above)
Comment (chleh) [inline Tests/BaseLib/TestFunctional.cpp]: Good point.

File: BaseLib/TMPUtil.h (hunk 2 above)
Comment (chleh) [inline BaseLib/TMPUtil.h]: As you said: Because it's C++14.

File: BaseLib/Functional.h (hunk 6 above)
Comment (endJunction) [inline BaseLib/Functional.h]:
```c++
std::remove_pointer<std::decay<std::string const* const>::type>::type
std::decay<std::remove_pointer<std::string const* const>::type>::type
```
are equal and return `std::string` w/o any qualifiers...

Comment (chleh) [inline BaseLib/Functional.h]: Since you obviously do some testing: Have you tested with the actual code of this PR? If I remove `std::remove_cv` I get the following error:

```
[ 71%] Building CXX object Tests/CMakeFiles/testrunner.dir/BaseLib/TestFunctional.cpp.o
src/Tests/BaseLib/TestFunctional.cpp: In member function ‘virtual void BaseLib_Functional_Test::TestBody()’:
src/Tests/BaseLib/TestFunctional.cpp:148:59: error: no matching function for call to ‘easyBind(double (A::*)() const, const A*&)’
         auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
                                                           ^
In file included from src/Tests/BaseLib/TestFunctional.cpp:13:0:
src/BaseLib/Functional.h:179:1: note: candidate: template<class Object, class MethodClass, class ReturnType, class ... Args> typename std::enable_if<std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<_Tp>::type>::type>::value, std::function<ReturnType(Args ...)> >::type BaseLib::easyBind(ReturnType (MethodClass::*)(Args ...), Object&&)
 easyBind(ReturnType (MethodClass::*method)(Args...), Object&& obj)
 ^
src/BaseLib/Functional.h:179:1: note:   template argument deduction/substitution failed:
src/Tests/BaseLib/TestFunctional.cpp:148:59: note:   types ‘ReturnType (MethodClass::)(Args ...)’ and ‘double (A::)() const’ have incompatible cv-qualifiers
         auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
                                                           ^
In file included from src/Tests/BaseLib/TestFunctional.cpp:13:0:
src/BaseLib/Functional.h:194:1: note: candidate: template<class Object, class MethodClass, class ReturnType, class ... Args> typename std::enable_if<std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<_Tp>::type>::type>::value, std::function<ReturnType(Args ...)> >::type BaseLib::easyBind(ReturnType (MethodClass::*)(Args ...) const, Object&&)
 easyBind(ReturnType (MethodClass::*method)(Args...) const, Object&& obj)
 ^
src/BaseLib/Functional.h:194:1: note:   template argument deduction/substitution failed:
src/BaseLib/Functional.h: In substitution of ‘template<class Object, class MethodClass, class ReturnType, class ... Args> typename std::enable_if<std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<_Tp>::type>::type>::value, std::function<ReturnType(Args ...)> >::type BaseLib::easyBind(ReturnType (MethodClass::*)(Args ...) const, Object&&) [with Object = const A*&; MethodClass = A; ReturnType = double; Args = {}]’:
src/Tests/BaseLib/TestFunctional.cpp:148:59:   required from here
src/BaseLib/Functional.h:194:1: error: no type named ‘type’ in ‘struct std::enable_if<false, std::function<double()> >’
src/BaseLib/Functional.h:206:55: note: candidate: template<class Object> typename BaseLib::detail::FunctionTraits<T>::FunctionType BaseLib::easyBind(Object&&)
 typename detail::FunctionTraits<Object>::FunctionType easyBind(Object&& obj)
                                                       ^
src/BaseLib/Functional.h:206:55: note:   template argument deduction/substitution failed:
src/Tests/BaseLib/TestFunctional.cpp:148:59: note:   candidate expects 1 argument, 2 provided
         auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
```

Comment (chleh) [inline BaseLib/Functional.h]: added a comment

Comment (endJunction) [inline BaseLib/Functional.h]: Thanks for clarification.

Just for general education: With clang-3.8.0 removing the first (around line 178) `remove_cv` is working fine, but removing the second one (around line 193) also causes a compilation error like:
```
[ 79%] Building CXX object Tests/CMakeFiles/testrunner.dir/BaseLib/TestFunctional.cpp.o
ogs/Tests/BaseLib/TestFunctional.cpp:148:24: error: no matching function for call to 'easyBind'
        auto fpc_get = BaseLib::easyBind(&A::getValue, apc);
                       ^~~~~~~~~~~~~~~~~
ogs/BaseLib/Functional.h:181:1: note: candidate template ignored: could not match 'type-parameter-0-2 (type-parameter-0-3...)' against 'double () const'
easyBind(ReturnType (MethodClass::*method)(Args...), Object&& obj)
^
ogs/BaseLib/Functional.h:192:5: note: candidate template ignored: disabled by 'enable_if' [with Object = const A *&, MethodClass = A, ReturnType = double, Args = <>]
    std::is_same<MethodClass, typename std::remove_pointer<typename std::decay<
    ^
ogs/BaseLib/Functional.h:207:55: note: candidate function template not viable: requires single argument 'obj', but 2 arguments were provided
typename detail::FunctionTraits<Object>::FunctionType easyBind(Object&& obj)
```

So maybe the tests are not testing all the combinations...

**Comment by endJunction** [comment]
:fast_forward:

**Comment by TomFischer** [comment]
:+1:
